### PR TITLE
Fix GC holes in array marshallers in the field scenario

### LINKED ISSUE #27491: Test failed: Interop\PInvoke\Array\MarshalArrayAsField\AsByValArray\AsByValArrayTest\AsByValArrayTest.cmd
**Job:**
[coreclr-gcstress-extra #20191026.1 Run Test Pri1 Windows_NT x64 checked](https://dev.azure.com/dnceng/public/_build/results?buildId=403511)

**Detail:**
https://helix.dot.net/api/2019-06-17/jobs/daa4e7dc-48d8-475e-a547-2e681715661d/workitems/Interop/console

**OS & Arch:**
windows x64

**Mode:**
COMPlus_TieredCompilation=0
COMPlus_GCStress=0xC
COMPlus_ReadyToRun=0
COMPlus_ZapDisable=1

**Note:**
when trying to repro this failure, I get the following result:
```
BEGIN EXECUTION
 "C:\Users\vbibu\403511_Tests_Windows_NT_x64_checked_gcstress-extra\Tests\Core_Root\corerun.exe" AsByValArrayTest.dll
RunTest1 : Marshal array as field as ByValArray in sequential struct as parameter.

TEST FAIL: Unable to load DLL 'MarshalArrayByValArrayNative' or one of its dependencies: The specified module could not be found. (0x8007007E)
Expected: 100
Actual: 101
END EXECUTION - FAILED
FAILED
```

**Log:**
```
Interop\PInvoke\Array\MarshalArrayAsField\AsByValArray\AsByValArrayTest\AsByValArrayTest.cmd [FAIL]
      
      Assert failure(PID 5840 [0x000016d0], Thread: 4896 [0x1320]): Consistency check failed: AV in clr at this callstack:
      ------
      CORECLR! OleVariant::MarshalNonBlittableRecordArrayOleToCom + 0x5BD (0x00007ffb`d3010b8d)
      CORECLR! OleVariant::MarshalRecordArrayOleToCom + 0x646 (0x00007ffb`d3013ed6)
      CORECLR! MngdFixedArrayMarshaler::ConvertContentsToManaged + 0x202 (0x00007ffb`d3194fb2)
      <no module>! <no symbol> + 0x0 (0x00007ffb`73879c3b)
      <no module>! <no symbol> + 0x0 (0x00000087`c5b9dcd0)
      <no module>! <no symbol> + 0x0 (0x00000087`c5b9d6b0)
      <no module>! <no symbol> + 0x0 (0x00000087`c5b9dcd0)
      <no module>! <no symbol> + 0x0 (0x000001b3`33dab0d8)
      <no module>! <no symbol> + 0x0 (0x00007ffb`73874860)
      <no module>! <no symbol> + 0x0 (0x00000087`c5b9d910)
      <no module>! <no symbol> + 0x0 (0x00007ffb`73bae6c0)
      <no module>! <no symbol> + 0x0 (0x00007ffb`73874860)
      <no module>! <no symbol> + 0x0 (0x00007ffb`73bbe752)
      <no module>! <no symbol> + 0x0 (0x00000001`00000000)
      <no module>! <no symbol> + 0x0 (0x00000024`00000000)
      <no module>! <no symbol> + 0x0 (0x00000000`00000064)
      <no module>! <no symbol> + 0x0 (0x00000087`c5b9d680)
      <no module>! <no symbol> + 0x0 (0x9abcdef0`12345678)
      <no module>! <no symbol> + 0x0 (0x000001b3`4bf84c60)
      -----
      .AV on tid=0x1320 (4896), cxr=00000087C5B9CA00, exr=00000087C5B9CEF0
      FAILED: false
      
      CORECLR! CHECK::Trigger + 0x2C5 (0x00007ffb`d2a8ce1d)
      CORECLR! CLRVectoredExceptionHandlerPhase3 + 0x364 (0x00007ffb`d2c19644)
      CORECLR! CLRVectoredExceptionHandlerPhase2 + 0x69 (0x00007ffb`d2c190c5)
      CORECLR! CLRVectoredExceptionHandler + 0x1C6 (0x00007ffb`d2c1903e)
      CORECLR! CLRVectoredExceptionHandlerShim + 0x17A (0x00007ffb`d2c1989a)
      NTDLL! RtlInitializeCriticalSection + 0x190 (0x00007ffc`01205e90)
      NTDLL! RtlImageNtHeaderEx + 0x17B (0x00007ffc`011dfa1b)
      NTDLL! KiUserExceptionDispatcher + 0x3A (0x00007ffc`012496fa)
      CORECLR! OleVariant::MarshalNonBlittableRecordArrayOleToCom + 0x5BD (0x00007ffb`d3010b8d)
      CORECLR! OleVariant::MarshalRecordArrayOleToCom + 0x646 (0x00007ffb`d3013ed6)
          File: F:\workspace\_work\1\s\src\vm\excep.cpp Line: 7537
          Image: C:\dotnetbuild\work\daa4e7dc-48d8-475e-a547-2e681715661d\Payload\CoreRun.exe
      
      
      Return code:      1
      Raw output file:      C:\dotnetbuild\work\daa4e7dc-48d8-475e-a547-2e681715661d\Work\d6a2823a-5b01-4fb2-8436-e3423017b592\Exec\Interop\PInvoke\Reports\Interop.PInvoke\Array\MarshalArrayAsField\AsByValArray\AsByValArrayTest\AsByValArrayTest.output.txt
      Raw output:
      BEGIN EXECUTION
       "C:\dotnetbuild\work\daa4e7dc-48d8-475e-a547-2e681715661d\Payload\corerun.exe" AsByValArrayTest.dll 
      RunTest1 : Marshal array as field as ByValArray in sequential struct as parameter.
      RunTest2 : Marshal array as field as ByValArray in sequential class as parameter.
      RunTest3 : Marshal array as field as ByValArray in explicit struct as parameter.
      RunTest4 : Marshal array as field as ByValArray in explicit class as parameter.
      RunTest5 : Marshal array as field as ByValArray in sequential class as return type.
      Expected: 100
      Actual: -1073740286
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
```

### LINKED ISSUE #27496: Test failed: Interop\ArrayMarshalling\ByValArray\MarshalArrayByValTest\MarshalArrayByValTest.cmd
**Job:**
[coreclr-gcstress0x3-gcstress0xc #20191027.1 Run Test Pri1 Windows_NT x86 checked](https://dev.azure.com/dnceng/public/_build/results?buildId=403705)

**Detail:**
https://helix.dot.net/api/2019-06-17/jobs/75ec302e-a8eb-458b-a2a4-2b4f05437c94/workitems/Interop/console

**OS & Arch:**
windows x86

**Mode:**
COMPlus_TieredCompilation=0
COMPlus_GCStress=0x3

**Note:**
I try to repro this failure, but get the following result:
```
BEGIN EXECUTION
 "C:\Users\vbibu\403705_Tests_Windows_NT_x86_checked_gcstress0x3-gcstress0xc\Tests\Core_Root\corerun.exe" MarshalArrayByValTest.dll
RunTest1 : Marshal array as field as ByValArray in sequential struct as parameter.
Unhandled exception. System.DllNotFoundException: Unable to load DLL 'MarshalArrayByValNative' or one of its dependencies: The specified module could not be found. (0x8007007E)
   at Test.TakeIntArraySeqStructByVal(S_INTArray_Seq s, Int32 size)
   at Test.RunTest1(String report) in F:\workspace\_work\1\s\tests\src\Interop\ArrayMarshalling\ByValArray\MarshalArrayByValTest.cs:line 745
   at Test.Main(String[] args) in F:\workspace\_work\1\s\tests\src\Interop\ArrayMarshalling\ByValArray\MarshalArrayByValTest.cs:line 1066
Expected: 100
Actual: -532462766
END EXECUTION - FAILED
FAILED
```

**Log:**
```
Interop\ArrayMarshalling\ByValArray\MarshalArrayByValTest\MarshalArrayByValTest.cmd [FAIL]
      
      Assert failure(PID 6052 [0x000017a4], Thread: 5388 [0x150c]): Consistency check failed: AV in clr at this callstack:
      ------
      CORECLR! Object::GetMethodTable + 0x3 (0x72b8c82a)
      CORECLR! ArrayBase::GetDataPtr + 0x1F (0x72ee2cfe)
      CORECLR! MngdFixedArrayMarshaler::ConvertContentsToNative + 0x21E (0x72fc972e)
      <no module>! <no symbol> + 0x0 (0x06fde900)
      <no module>! <no symbol> + 0x0 (0x06fde771)
      <no module>! <no symbol> + 0x0 (0x06fdd84c)
      <no module>! <no symbol> + 0x0 (0x06fd4139)
      CORECLR! CallDescrWorkerInternal + 0x34 (0x72bf80d1)
      CORECLR! CallDescrWorker + 0xD7 (0x72dd41e7)
      CORECLR! CallDescrWorkerWithHandler + 0x11B (0x72dd4364)
      CORECLR! MethodDescCallSite::CallTargetWorker + 0x5F9 (0x72dd4993)
      CORECLR! RunMainInternal + 0x1A4 (0x72dce2c6)
      CORECLR! ``RunMain'::`29'::__Body::Run'::`5'::__Body::Run + 0x42 (0x72dcdeb6)
      CORECLR! `RunMain'::`29'::__Body::Run + 0x5F (0x72dcdf68)
      CORECLR! RunMain + 0xD1 (0x72dce0a0)
      CORECLR! Assembly::ExecuteMainMethod + 0x1A0 (0x72dcb34b)
      CORECLR! CorHost2::ExecuteAssembly + 0x409 (0x72b85489)
      CORERUN! TryRun + 0x6C1 (0x00b9d0dd)
      CORERUN! wmain + 0x98 (0x00b9dd06)
      -----
      .AV on tid=0x150c (5388), cxr=0077D2A8, exr=0077D258
      FAILED: false
      
      CORECLR! CHECK::Trigger + 0x2FB (0x72b57ffe)
      CORECLR! CLRVectoredExceptionHandlerPhase3 + 0x25E (0x72c44202)
      CORECLR! CLRVectoredExceptionHandlerPhase2 + 0x4E (0x72c43e71)
      CORECLR! CLRVectoredExceptionHandler + 0x13C (0x72c43e17)
      CORECLR! CLRVectoredExceptionHandlerShim + 0x168 (0x72c44438)
      NTDLL! LdrSetAppCompatDllRedirectionCallback + 0x1CB63 (0x779b04c3)
      NTDLL! RtlUnwind + 0x1BA (0x7797051a)
      NTDLL! KiUserExceptionDispatcher + 0xF (0x77980fff)
      CORECLR! MngdFixedArrayMarshaler::ConvertContentsToNative + 0x21E (0x72fc972e)
      <no module>! <no symbol> + 0x0 (0x06fde900)
          File: F:\workspace\_work\1\s\src\vm\excep.cpp Line: 7537
          Image: C:\dotnetbuild\work\75ec302e-a8eb-458b-a2a4-2b4f05437c94\Payload\CoreRun.exe
      
      
      Return code:      1
      Raw output file:      C:\dotnetbuild\work\75ec302e-a8eb-458b-a2a4-2b4f05437c94\Work\b2113a47-cddc-48ba-be26-f2e286242216\Exec\Interop\ArrayMarshalling\Reports\Interop.ArrayMarshalling\ByValArray\MarshalArrayByValTest\MarshalArrayByValTest.output.txt
      Raw output:
      BEGIN EXECUTION
       "C:\dotnetbuild\work\75ec302e-a8eb-458b-a2a4-2b4f05437c94\Payload\corerun.exe" MarshalArrayByValTest.dll 
      RunTest1 : Marshal array as field as ByValArray in sequential struct as parameter.
      Expected: 100
      Actual: -1073740286
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
```

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Correctly protect references to objects in manually-managed code called by the array IL marshalers in the field scenarios. Unlike the non-field scenarios where the "managed home" address is a local variable on the stack, the "managed home" address for field scenarios is a byref into the managed object. If the managed object is a layout-class, it can be relocated/collected.

Also, use RuntimeHelpers.GetRawData to simplify layout class marshalling and reduce weird IL patterns in the marshalers when there are C# methods with better readability that we can reuse.

Fixes #27496 
Fixes #27491 
Fixes dotnet/corefx#42201

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky Is there a test we can add that would have caught this issue when running under GC stress?

**Comment by jkoritzinsky** [comment]
Well the two fixed-array marshalling test projects did catch this under GCStress. That's what #27496 and #27491 are from. This might not show up on tier0 code since (at least the specific failure in #27496) only shows up when the gcinfo for an object is as strict (shortest lifetime) as possible.

**Comment by jkoritzinsky** [comment]
I initially missed this since I copied the pattern of the array marshalers from the ones already in the code base and didn't realize that the differences in the scenario (param/retval vs field) were enough to possibly cause GC holes.

**Comment by jkoritzinsky** [comment]
@trylek I can't run the GCStress pipeline to validate this in CI. I'm getting the error:
> Job run_test_p1__Linux_x64_checked depends on unknown job build_test_p1_OSX_x64_checked.

Looks like your recent changes to the pipeline may have broken the jobs that don't include an OSX leg.

**Comment by trylek** [comment]
Hmm, that's unfortunate, apologies about that. I think this should be fixable by tweaking the conditions in platform-matrix - it's true that with my change any Linux test execution looks for managed tests components built on OSX. I have a meeting in 5 minutes, I'll look into fixing that right after we finish. Is this the gcstress-extra pipeline where you're hitting this? I definitely need to know how to test the fix.

File: src/vm/ilmarshalers.cpp [hunk 1]
```diff
@@ -4669,6 +4680,8 @@ FCIMPL3(void, MngdFixedArrayMarshaler::ConvertSpaceToManaged, MngdFixedArrayMars
     //
     SetObjectReference(pManagedHome, AllocateSzArray(pThis->m_Array, pThis->m_cElements));
```
Comment (jkotas) [inline src/vm/ilmarshalers.cpp]: This is potential GC hole. It is valid for C++ compiler to fetch the value of `pManagedHome`, then call AllocateSzArray - that will move the object that pManagedHome points to, and then call SetObjectReference - with the bogus value of pManagedHome.

This needs to be split into multiple lines

**Comment by jkoritzinsky** [comment]
@trylek It's the gcstress_0x3_0xc pipeline.

**Comment by jkotas** [comment]
```
Assert failure(PID 3592 [0x00000e08], Thread: 4268 [0x10ac]): Consistency check failed: AV in clr at this callstack:
------
CORECLR! OBJECTREF::OBJECTREF + 0x10 (0x00007ffb`8ccb0460)
CORECLR! MngdFixedArrayMarshaler::ClearNativeContents + 0xCA (0x00007ffb`8d09185a)
```

**Comment by jkoritzinsky** [comment]
I'll take a look when I get into the office tomorrow.

File: src/vm/ilmarshalers.cpp [hunk 2]
```diff
@@ -4937,17 +4962,20 @@ FCIMPL3(void, MngdSafeArrayMarshaler::ConvertSpaceToManaged, MngdSafeArrayMarsha
                 COMPlusThrow(kSafeArrayRankMismatchException, IDS_EE_SAFEARRAYSZARRAYMISMATCH);
             }
         }
-    
-        SetObjectReference(pManagedHome,
-            (OBJECTREF) OleVariant::CreateArrayRefForSafeArray((SAFEARRAY*) *pNativeHome,
+
+        OBJECTREF arrayRef = (OBJECTREF) OleVariant::CreateArrayRefForSafeArray((SAFEARRAY*) *pNativeHome,
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.cpp]: >(SAFEARRAY*) *pNativeHome [](start = 80, length = 25)

Can you do this once above? It is done in a bunch of places in this branch.

**Comment by jkoritzinsky** [comment]
Failures in the GCStress pipeline are due to the following issues:

- Linux_arm timeouts which have been happening for a few weeks on and off.
-  crossgen2smoke, which is being tracked by #27497
- pinvoke_examples, which is being tracked by #27602

**Pull request merged**